File: Wordle/Regina.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace Wordle
{
    public class Regina : IWordleBot
    {
        public List<GuessResult> Guesses { get; set; }

        private List<string> options = new List<string>();

        public Regina()
        {
            Guesses = new List<GuessResult>();

            // read all 5 letter words into options
            try
            {
                using (StreamReader streamReader = new StreamReader("../../../data/five_letter_words.txt"))
                {
                    string line;
                    while ((line = streamReader.ReadLine()) != null)
                    {
                        line = line.Trim();
                        options.Add(line);
                        continue;
                    }
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
                return;
            }

            //options.Sort();
        }

        public string GenerateGuess()
        {
            if (Guesses.Count == 0)
            {
                return "crane";
            }

            GuessResult lastGuess = Guesses[^1];
            Regex rgx = GenerateRegex(lastGuess);
            Regex keepers = GenerateKeeperSet();

            // eliminate impossible solutions
            foreach (string word in options.ToList())
            {
                if (!(rgx.IsMatch(word) && keepers.IsMatch(word)))
                {
                    options.Remove(word);
                }
            }

            if (Guesses.Count == 1)
            {
                return "toils";
            }

            return options[0];
        }

        public string GenerateEmojiString()
        {
            string emoji = $"Regina | {DateTime.Today.ToShortDateString()}";

            foreach (GuessResult gr in Guesses)
            {
                emoji += "\n";

                foreach (LetterGuess lg in gr.Guess)
                {
                    switch (lg.LetterResult)
                    {
                        case LetterResult.Correct:
                            emoji += "+";
                            break;
                        case LetterResult.Misplaced:
                            emoji += "o";
                            break;
                        case LetterResult.Incorrect:
                            emoji += "-";
                            break;
                        default:
                            break;
                    }
                }
            }

            return emoji;
        }

        private Regex GenerateRegex(GuessResult guess)
        {
            string pattern = "";

            foreach (LetterGuess lg in guess.Guess)
            {
                if (lg.LetterResult == LetterResult.Correct)
                {
                    pattern += $@"{lg.Letter}";
                }
                else
                {
                    pattern += $@"[^{lg.Letter}]";
                }
            }

            return new Regex(pattern);
        }

        private Regex GenerateKeeperSet()
        {
            string keepers = "";

            foreach (GuessResult gr in Guesses)
            {
                foreach (LetterGuess lg in gr.Guess)
                {
                    if (lg.LetterResult != LetterResult.Incorrect && !keepers.Contains(lg.Letter))
                    {
                        keepers += $@"(?=.*{lg.Letter})";
                    }
                }
            }

            return new Regex(keepers);
        }
    }
}

File: pywordle/regina.py
"""
Regina, the Wordle bot that uses regex to (almost) always win
"""
import re
from letter_guess import LetterGuess, LetterResult
from guess_result import GuessResult


class Regina:
    def __init__(self):
        self.guesses = []
        self._options = []

        # read all 5 letter words into options
        for word in open("./words.txt"):
            word = word.strip()
            if len(word) == 5:
                self._options.append(word)

    def generate_guess(self) -> str:
        if len(self.guesses) == 0:
            return "crane"

        last_guess = self.guesses[-1]
        rgx = self._generate_regex(last_guess)
        keepers = self._generate_keeper_set()

        # eliminate impossible solutions
        for word in list(self._options):
            if not (re.match(rgx, word) and re.match(keepers, word)):
                self._options.remove(word)

        if len(self.guesses) == 1:
            return "toils"

        return self._options[0]

    def _generate_regex(self, guess: GuessResult):
        pattern = ""

        for lg in guess.guess:
            if lg.letter_result == LetterResult.correct:
                pattern += lg.letter
            else:
                pattern += f"[^{lg.letter}]"

        return pattern

    def _generate_keeper_set(self):
        keepers = ""

        for gr in self.guesses:
            for lg in gr.guess:
                if lg.letter_result != LetterResult.incorrect and not keepers.__contains__(lg.letter):
                    keepers += f"(?=.*{lg.letter})"

        return keepers

    def __str__(self):
        return "Hi, I'm Regina, a Wordle-obsessed bot. Bet you can't solve today's word in fewer guesses than me."
